### PR TITLE
store/s3: try to guess size to enable multipart uploads

### PR DESCRIPTION
**Pull request by adrien-f**
Signed-off-by: Adrien Fillon <adrien.fillon@cdiscount.com>

I noticed an issue where a S3 multipart upload would fail with a `413 Entity Too Large` error. After investigating, I noticed we were not sending the size of the buffer/chunk/file/... to the minio-go library and that would cause the multipart upload to be a single part upload.

From what I've seen, the `Upload` method is only called with a `os.File` object : https://github.com/improbable-eng/thanos/blob/a09d41dd6534f28365eede43fd024fed3dcc0750/pkg/objstore/objstore.go#L80

I'm currently not satisfied with what I did but it works :shipit: ! I open this PR so we can figure out the best way to handle this. In case of real buffers, to compute their size I would need to duplicate it in memory from what I read.

## Changes

* s3: Send correct file size to support multipart uploads

## Verification

* Deployed the patched binary on our production clusters, uploads started again.

For your information, we use Ceph as our S3 provider.

### CODE REVIEW

**Comment by bwplotka** [comment]
I think we had some discussion offline on this. Definitely this PR is wrong as it will panic if io.Reader is not os.File.

So we have 3 options:
- extend interface with sizeBytes int64 argument.
- try to cast reader into os.File safely. 
- ask minio maintainers to not make any assumptions based on size - the same way as GCS implementation does. 

As the 3rd option is most likely long and potentially not possible due to s3 API limitations and taking into consideration fact that in most cases we have os.File passed as reader I think option 1 makes sense... 

Option 2 is too hidden and it makes assumptions that are easy to miss.

Thoughts @domgreen?

**Comment by adrien-f** [comment]
I went with option 1. I guess that should help with #467 @jojohappy ?

File: pkg/objstore/inmem/inmem.go [hunk 1]
```diff
@@ -124,7 +124,7 @@ func (b *Bucket) Exists(_ context.Context, name string) (bool, error) {
 }
 
 // Upload writes the file specified in src to into the memory.
-func (b *Bucket) Upload(_ context.Context, name string, r io.Reader) error {
+func (b *Bucket) Upload(_ context.Context, name string, r io.Reader, sizeBytes int64) error {
```
Comment (bwplotka) [inline pkg/objstore/inmem/inmem.go]: same here

File: pkg/objstore/gcs/gcs.go [hunk 2]
```diff
@@ -124,7 +124,7 @@ func (b *Bucket) Exists(ctx context.Context, name string) (bool, error) {
 }
 
 // Upload writes the file specified in src to remote GCS location specified as target.
-func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
+func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader, sizeBytes int64) error {
```
Comment (bwplotka) [inline pkg/objstore/gcs/gcs.go]: I am pretty sure we can reuse this knowledge, but you can leave that to others. Just please use _ for arg name that is not used.

File: pkg/objstore/azure/azure.go [hunk 3]
```diff
@@ -211,7 +211,7 @@ func (b *Bucket) Exists(ctx context.Context, name string) (bool, error) {
 }
 
 // Upload the contents of the reader as an object into the bucket.
-func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
+func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader, sizeBytes int64) error {
```
Comment (bwplotka) [inline pkg/objstore/azure/azure.go]: @vglafirov is this something we can use in azure?

If not @adrien-f let's make it `_` as well

**Comment by jojohappy** [comment]
@bwplotka sorry for my poor English, `I` means the consumer, and IMO, it should not give the size of `Reader` to the object storage client particularly, I mean I like the 3rd option.

**Comment by bwplotka** [comment]
Well but for COS you need size as well? Otherwise it does not work either: https://github.com/improbable-eng/thanos/pull/467/files#r209909773

**Comment by jojohappy** [comment]
Yes, actually the 1st option will solve the problem now and make sence. I accept it and will ask the cos-client maintainer to not make the size sensitively.

File: pkg/objstore/objstore.go [hunk 4]
```diff
@@ -83,7 +83,14 @@ func UploadFile(ctx context.Context, logger log.Logger, bkt Bucket, src, dst str
 	}
 	defer runutil.CloseWithLogOnErr(logger, r, "close file %s", src)
 
-	if err := bkt.Upload(ctx, dst, r); err != nil {
+	fileSize := int64(-1)
```
Comment (brancz) [inline pkg/objstore/objstore.go]: Maybe I'm missing something, but this initialization doesn't seem to be necessary as the `Stat` call must succeed.

**Comment by domgreen** [comment]
The COS implementation does this here https://github.com/improbable-eng/thanos/pull/467/files#r209909773 
I tend to agree not extending the interface feels better and therefore only the needed implementations do this extra step. Also from my quick look type assertion does not seem to be overly expensive in Go.

**Comment by bwplotka** [comment]
It's not about cost of type assertion - it's marginal. 

It's about fact that provider has different logic or even fails if the passed io.Reader is NOT os.File. The fact that implmentation assumes `os.File` is wrong as our interface contract does not mention it. We expect user to pass any `io.Reader`.

So in future someone might add usage that will wrap `*os.File` or like we do now pass `bytes.Buffer` (like we do in unit tests). This is literally abstraction leak, as implementation poses limitation that cause unexpected side effects (multipart disabled for mino or failure for COS).

More details why this problem can be categorized as abstraction leak: https://medium.com/@fagnerbrack/repairing-the-leaky-abstraction-e726baab91b5

So there is no other way than either:
1. expose leak/limitation and be open about this in form of passing the size. Mentioning that `io.Reader` has to be `os.File` in docs is clearly not enough (also we don't need whole file, just `size int64`, and we want abstraction to be as thin as possible)
2. Guess size by doing buffering as @brancz suggested. Quite complex for us to do it on top of minio or COS, but doable. Worth to discuss this with authors of packages we use for problematic providers.

For example, `GCS` package we use is doing 2, that's why it does not need preknown size.

**Comment by bwplotka** [comment]
any friday thoughts @domgreen @brancz ^^

**Comment by stefan-improbable** [comment]
@bwplotka @adrien-f can we change the interface to `Upload(ctx context.Context, name string, r os.File) error` or do we really need that `io.Reader` there?

**Comment by bwplotka** [comment]
@stefan-improbable Why is that better? Also passing `os.File` would be violation of idiomatic "Accept interfaces, return struct" guideline. Well explained with actually similar example here: http://idiomaticgo.com/post/best-practice/accept-interfaces-return-structs/

Also, implementation does not need `os.File`  just `io.Reader` and size. So rather `ReaderSize` interface would be something slightly better here, but it would be more difficult to use than `io.Reader` and just single int size at the end.

**Comment by brancz** [comment]
I think we should consult the minio people, as to me it seems that this should be possible without a change in the interface, looking at their internal handling, this should actually work, as it should attempt to perform multi part on a best effort basis: https://github.com/minio/minio-go/blob/70799fe8dae6ecfb6c7d7e9e048fce27f23a1992/api-put-object.go#L157

**Comment by adrien-f** [comment]
> I think we should consult the minio people, as to me it seems that this should be possible without a change in the interface, looking at their internal handling, this should actually work, as it should attempt to perform multi part on a best effort basis: https://github.com/minio/minio-go/blob/70799fe8dae6ecfb6c7d7e9e048fce27f23a1992/api-put-object.go#L157
> 
> (mind you I'm well jetlagged right now so may very well be missing something)

While digging this issue, I looked into this piece of code I got to https://github.com/minio/minio-go/blob/70799fe8dae6ecfb6c7d7e9e048fce27f23a1992/api-put-object.go#L184 then https://github.com/minio/minio-go/blob/2508a98a820e993b0307bbac4db9731e4e6f89a0/api-put-object-common.go#L72 that goes with this:

```
        // object size is '-1' set it to 5TiB.
	if objectSize == -1 {
		objectSize = maxMultipartPutObjectSize
	}
```

The following issue https://github.com/minio/minio-go/issues/848#issuecomment-338082471 propose an alternative solution that involves manually uploading the blob parts by parts. Adding such a relative complexity to Thanos felt unneeded.

**Comment by stefan-improbable** [comment]
@bwplotka
> @stefan-improbable Why is that better? Also passing os.File would be violation of idiomatic "Accept interfaces, return struct" guideline. Well explained with actually similar example here: http://idiomaticgo.com/post/best-practice/accept-interfaces-return-structs/

Passing reader and length seems a bit weird, but I don't have any other argument. Possibly just that all callers will have to calculate size as opposed to it being calculated in the one AWS-specific function. As for the accept-interfaces-and-return-structs, you can open the file for reading/writing only and test it using ioutil.TempFile.
But in the end, the reader+size is not too bad and solves our current problem, so I'm happy to go with it.

**Comment by adrien-f** [comment]
Any last minute changes I should make ?

File: pkg/objstore/objstore.go [hunk 5]
```diff
@@ -22,7 +22,7 @@ type Bucket interface {
 	BucketReader
 
 	// Upload the contents of the reader as an object into the bucket.
-	Upload(ctx context.Context, name string, r io.Reader) error
+	Upload(ctx context.Context, name string, r io.Reader, sizeBytes int64) error
```
Comment (bwplotka) [inline pkg/objstore/objstore.go]: Let's comment properly here, why this has size and reader interface. Worth to mention PR https://github.com/improbable-eng/thanos/pull/617 as we have reference discussion why etc.

File: pkg/objstore/objstore.go (hunk 4 above)
Comment (bwplotka) [reply pkg/objstore/objstore.go]: Yes, let's remove it

File: pkg/objstore/objstore.go (hunk 5 above)
Comment (bwplotka) [reply pkg/objstore/objstore.go]: I would also add a comment saying `TODO(PR #617): Consider removing requirement on pre-known length when all providers will support this.`

**Comment by bwplotka** [comment]
@adrien-f Can you address small comments and change title/commit title to exclude `WIP` prefix

**Comment by bwplotka** [comment]
@adrien-f And we need CHANGELOG maybe as this loosens our requirements towards client provider implementations (if someone wants to add one)?

**Comment by domgreen** [comment]
Does the update of the COS API effect this discussion? https://github.com/improbable-eng/thanos/issues/679 

I am happy to add the `int` to the interface to unblock and gain perf improvements however, think that we may have to revisit the interface in the future so it is more future proof.

**Comment by domgreen** [comment]
Re `CHANGELOG` ... I'm not sure we want to update as this is an internal cahnge that will not effect users really only provider devs.

To further followup on my previous comment now it is only minio that needs the size ... maybe we treat this as the special case and do the casting? Sorry to keep throwing the cat amongst the pigeons.

**Comment by jojohappy** [comment]
I have make a test and confirmed that the new version of COS provider had been supported uploading files without the length.

**Comment by adrien-f** [comment]
I made the last requested changes I think.
I'll be more than happy to help work on the transition to go-cloud :smile: 
Thanks everyone for the review and the feedback, I realize this change isn't the best solution, and nothing is more frustrating than this situation. I hope to be able to resolve this soon durably.

**Comment by bwplotka** [comment]
Oh boy, COS news means that this interface change is indeed weird.

Thanks for doing all this work @adrien-f and I feel like in this case casting to os.File and warn if we cannot do it so `multipart upload` will be disabled (and mention https://github.com/improbable-eng/thanos/issues/678) is proper direction here.

Thanks all for valuable discussion! (:

File: pkg/objstore/s3/s3.go [hunk 6]
```diff
@@ -250,7 +250,14 @@ func (b *Bucket) Exists(ctx context.Context, name string) (bool, error) {
 
 // Upload the contents of the reader as an object into the bucket.
 func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
-	_, err := b.client.PutObjectWithContext(ctx, b.name, name, r, -1,
+	// TODO(PR #617): Consider removing requirement on pre-known length when all providers will support this.
+	fileInfo, err := r.(*os.File).Stat()
+	if err != nil {
+		return errors.Wrapf(err, "could not stat local file %s", name)
```
Comment (bwplotka) [inline pkg/objstore/s3/s3.go]: Well, we agreed for best effort logic, so we need warning instead I think, right?

Comment (adrien-f) [reply pkg/objstore/s3/s3.go]: My bad. I really need holidays :face_with_head_bandage:

**Pull request merged**